File: Data Structures & Algorithms/rotate-matrix/submission-1.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        int m = n % 2 == 0 ? n / 2 : (n - 1) / 2;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < (n + 1) / 2; j++) {
                System.out.println(matrix[i][j]);
                int temp = matrix[i][j];
                matrix[i][j] = matrix[n-j-1][i];
                matrix[n-j-1][i] = matrix[n-i-1][n-j-1];
                matrix[n-i-1][n-j-1] = matrix[j][n-i-1];
                matrix[j][n-i-1] = temp;
            }
        }
    }
}

File: Data Structures & Algorithms/spiral-matrix/submission-1.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        direction = [(0,1), (1,0), (0,-1),(-1,0)]
        answer = []
        m = len(matrix)
        n = len(matrix[0])
        i = 0 
        j = 0
        d = 0
        while len(answer) < m * n:
            answer.append(matrix[i][j])
            matrix[i][j] = None

            ni = i + direction[d][0]
            nj = j + direction[d][1]
            if ni < 0 or ni >= m or nj <0 or nj >= n or matrix[ni][nj] is None:
                d += 1
                d %=4
            i += direction[d][0]
            j += direction[d][1]
        return answer